src/main/java/com/google/devtools/build/lib/util/SingleEntryWeakLoadingCache.java

```diff
@@ -0,0 +1,48 @@
+package com.google.devtools.build.lib.util;
+
+import com.google.devtools.build.lib.concurrent.ThreadSafety;
+import java.lang.ref.WeakReference;
+import java.util.AbstractMap;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * An identity-based cache that weakly holds a single entry.
+ *
+ * <p>This is useful to associate data with {@link com.google.devtools.build.skyframe.SkyValue}s
+ * that are effectively singletons.
+ */
+@ThreadSafety.ThreadSafe
+public final class SingleEntryWeakLoadingCache<K, V> {
+  private final Function<K, V> loader;
+  @Nullable private WeakReference<AbstractMap.SimpleImmutableEntry<K, V>> weakEntry;
```

Comment:
There is no other reference to the `SimpleImmutableEntry`, won't it be collected on every GC?

Reply:
As an alternative, forget about trying to make it weak, and just "clear" when a new key is seen. That can probably be inlined into `StarlarkCustomCommandLine` instead of being its own class?

Reply:
Good catch, I guess this would need separate key and value fields. I implemented your suggestion instead.

Reply:
Here's how I would go about it: https://bazel-review.googlesource.com/c/bazel/+/259781. Figured it was faster than going back and forth over review.

Reply:
Thanks for taking care of all the details, looks good to me! Did you see something in the profile that nudged you towards two separate interners even though the bounded cache overhead should be gone?

Reply:
Wasn't influenced by profiles so much as I thought it was easier to understand - conceptually, we only need `StarlarkSemantics` when `map_each` is present. So a regular old interner is good for all other cases. Also felt that this logic was easier to understand "inline" rather than behind an abstraction like `InternerCache`.

I'll approve and import once you push my suggested changes.

Reply:
Also fyi, looking at that profile from yesterday gave me the idea to mail https://bazel-review.googlesource.com/c/bazel/+/259743.